File: component_hotfix/src/main/java/com/link/component_hotfix/TinkerService.java
package com.link.component_hotfix;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.Environment;
import android.os.Handler;
import android.os.IBinder;
import android.os.Message;

import java.io.File;

public class TinkerService extends Service {
    private String mPatchFileDir;
    private String mPatchFile;
    private final int CHECK_PATCH_UPDATE = 0x01;
    private final int DOWNLOAD_PATH = 0x02;
    private final String PATH_DOT = ".apk"; //补丁文件后缀

    private Handler mHandler = new Handler() {
        @Override
        public void handleMessage(Message msg) {
            switch (msg.what) {
                case CHECK_PATCH_UPDATE:
                    checkPatchUpdate();
                    break;
                case DOWNLOAD_PATH:
                    downPatch();
                    break;
                default:
                    break;
            }
        }
    };

    //对外提供启动servcie方法
    public static void runTinkerService(Context context) {
        try {
            Intent intent = new Intent(context, TinkerService.class);
            context.startService(intent);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    //用来与被启动者通信的接口
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        init(); //初始化Patch文件目录
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        mHandler.sendEmptyMessage(CHECK_PATCH_UPDATE); //检查是否有补丁
        return START_NOT_STICKY; //START_NOT_STICKY表示service如果被系统回收不重启
    }

    private void checkPatchUpdate() {

//        RetrofitClient.getInstance()
//                .create(HotfixHttpService.class)
//                .getRepireVersionInfo(CommonUtil.getPackageInfo().versionName)
//                .compose(RxUtils.schedulersTransformer())
//                .subscribe((Consumer<HotFixBase>) hotFixBase -> {
//                    if (hotFixBase.getCode() == 1 && hotFixBase.getList() != null && hotFixBase.getList().size() > 0) {
//                        mAppVersion = hotFixBase.getList().get(0);
//                        downPatch(mAppVersion);
//                    }
//                }, throwable -> stopSelf());


    }


    private void downPatch() {
//        //保存的补丁文件命名规则：xxx/1.0.0.1.apk 1.0.0是正常版本 .1是补丁版本号
//        if (null != appVersion) {
//            mPatchFile = mPatchFileDir.concat(appVersion.getVersionCode()).concat(PATH_DOT);
//            //判断该补丁是否存在，如果已经存在，不需要重复下载
//            File patch = new File(mPatchFile);
//            if (patch.exists()) {
//                return;
//            }
//            OkGo.<File>get(appVersion.getDownloadUrl())
//                    .tag(this)
//                    .execute(new FileCallback(mPatchFileDir, appVersion.getVersionCode().concat(PATH_DOT)) {
//                        @Override
//                        public void onStart(Request<File, ? extends Request> request) {
//                            super.onStart(request);
//                        }
//
//                        @Override
//                        public void downloadProgress(Progress progress) {
//
//                        }
//
//                        @Override
//                        public void onSuccess(Response<File> response) {
//                            TinkerManager.loadPatch(mPatchFile);
//                            stopSelf();
//                        }
//
//                        @Override
//                        public void onError(Response<File> response) {
//                            super.onError(response);
//                            stopSelf();
//                        }
//                    });
//
//        }
    }


    public void init() {
        if (Environment.getExternalStorageState().equals(Environment.MEDIA_MOUNTED)) {
            //sd卡已挂载
            mPatchFileDir = getExternalCacheDir().getAbsolutePath() + "/patch/";
            File patchDir = new File(mPatchFileDir);
            try {
                if (!patchDir.exists()) {
                    patchDir.mkdir();
                }
            } catch (Exception e) {
                e.printStackTrace();
                stopSelf();
            }
        } else {
//            ToastUtils.showShortSafe("SD卡未挂载");
            stopSelf();
        }
    }

}

File: app/src/main/java/com/link/fan/tinker/TinkerService.kt
package com.link.fan.tinker

import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.Environment
import android.os.Handler
import android.os.IBinder
import android.os.Message
import android.util.Log

import java.io.File

import cn.bmob.v3.BmobQuery
import cn.bmob.v3.exception.BmobException
import cn.bmob.v3.listener.DownloadFileListener
import cn.bmob.v3.listener.FindListener
import com.link.librarymodule.utils.CommonUtil
/**
 * <pre>
 *  copyright:TS
 *  author:wujia
 *  create:2019-10-26-09:46
 *  email:wujia0916@thundersoft.com
 *  description:
 * <pre>
 */
class TinkerService : Service() {
    //补丁的文件夹
    private var mPatchFileDir: String? = null
    //补丁的具体路径
    private var mPatchFilePath: String? = null
    //补丁文件后缀
    private val PATH_DOT = ".apk"

    private val CHECK_PATCH_UPDATE = 0x01
    private val DOWNLOAD_PATH = 0x02

    private val mHandler = object : Handler() {
        override fun handleMessage(msg: Message) {
            when (msg.what) {
                CHECK_PATCH_UPDATE -> checkPatchUpdate()
                DOWNLOAD_PATH -> downPatch()
                else -> {
                }
            }
        }
    }

    //用来与被启动者通信的接口
    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    override fun onCreate() {
        super.onCreate()
        init() //初始化Patch文件目录
    }

    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        mHandler.sendEmptyMessage(CHECK_PATCH_UPDATE) //检查是否有补丁
        return START_NOT_STICKY //START_NOT_STICKY表示service如果被系统回收不重启
    }

    private var mHotfixEntity: HotfixEntity? = null

    private fun checkPatchUpdate() {
        val query = BmobQuery<HotfixEntity>()
        query.addWhereEqualTo("versionName", CommonUtil.getPackageInfo().versionName)
        query.findObjects(object : FindListener<HotfixEntity>() {
            override fun done(data: MutableList<HotfixEntity>?, e: BmobException?) {
                if (e == null) {
                    if (data != null) {
                        mHotfixEntity = data[0]
                        mHandler.sendEmptyMessage(DOWNLOAD_PATH) //检查是否有补丁
                    }else{
                        stopSelf()
                    }
                } else {
                    stopSelf()
                }
            }

        })
    }


    private fun downPatch() {
        if (mHotfixEntity == null) {
            return
        }
        mPatchFilePath = mPatchFileDir + mHotfixEntity!!.patchCode + PATH_DOT
        //判断该补丁是否存在，如果已经存在，不需要重复下载
        val patch = File(mPatchFilePath)
        if (patch.exists()) {
            return
        }
        mHotfixEntity!!.patch.download(patch, object : DownloadFileListener() {
            override fun onProgress(p0: Int?, p1: Long) {

            }

            override fun done(savePath: String?, p1: BmobException?) {
                Log.e("TAG", savePath + "-")
                if (!savePath.isNullOrEmpty()) {
                    TinkerManager.loadPatch(savePath)
                }
                stopSelf()
            }

            override fun doneError(code: Int, msg: String?) {
                super.doneError(code, msg)
                stopSelf()
            }

        })

    }


    private fun init() {
        if (Environment.getExternalStorageState() == Environment.MEDIA_MOUNTED) {
            //sd卡已挂载
            mPatchFileDir = externalCacheDir!!.absolutePath + "/patch/"
            val patchDir = File(mPatchFileDir)
            try {
                if (!patchDir.exists()) {
                    patchDir.mkdir()
                }
            } catch (e: Exception) {
                e.printStackTrace()
                stopSelf()
            }

        } else {
            //ToastUtils.showShortSafe("SD卡未挂载");
            stopSelf()
        }
    }

    companion object {

        //对外提供启动servcie方法
        fun runTinkerService(context: Context) {
            try {
                val intent = Intent(context, TinkerService::class.java)
                context.startService(intent)
            } catch (e: Exception) {
                e.printStackTrace()
            }

        }
    }

}
